### PR TITLE
RDM-3288 Implement external Start new case endpoint in API v2

### PR DESCRIPTION
**Pull request by andrewfolga**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RDM-3288

### Change description ###
Implement external Start new case endpoint in API v2:
- Produces custom vnd media type
- Use HATEOAS/HAL resource
- Swagger doc annotation
- Functional tests

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/ccd/domain/service/startevent/DefaultStartEventOperation.java [hunk 1]
```diff
@@ -115,31 +136,30 @@ public StartEventTrigger triggerStartForDraft(final String uid,
...
+                                                     final CaseType caseType,
                                                      final String eventTriggerId,
                                                      final Boolean ignoreWarning,
                                                      final Supplier<CaseDetails> caseDetailsSupplier) {
-        final CaseType caseType = getCaseType(caseTypeId);
-        final CaseEvent eventTrigger = getEventTrigger(caseTypeId, eventTriggerId, caseType);
+        final CaseEvent eventTrigger = getEventTrigger(caseType.getId(), eventTriggerId, caseType);
 
-        validateJurisdiction(jurisdictionId, caseTypeId, caseType);
+        validateJurisdiction(caseType.getJurisdictionId(), caseType.getId(), caseType);
```
Comment (vlaurin) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/startevent/DefaultStartEventOperation.java]: I don't think this is required anymore. If the jurisdiction is always taken from the case type, then validating it is not required.

File: src/main/java/uk/gov/hmcts/ccd/domain/service/startevent/DefaultStartEventOperation.java [hunk 2]
```diff
@@ -115,31 +136,30 @@ public StartEventTrigger triggerStartForDraft(final String uid,
...
                                       eventTriggerId,
                                       ignoreWarning,
-            () -> getDraftDetails(draftReference));
+                                      () -> getDraftDetails(draftReference));
     }
 
-    private StartEventTrigger buildStartEventTrigger(final String uid,
-                                                     final String jurisdictionId,
-                                                     final String caseTypeId,
+    private StartEventTrigger buildStartEventTrigger(final Supplier<String> uidSupplier,
```
Comment (vlaurin) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/startevent/DefaultStartEventOperation.java]: Why is a supplier needed for the `uid`?

File: src/main/java/uk/gov/hmcts/ccd/v2/V2.java [hunk 3]
```diff
@@ -11,5 +11,9 @@
 
         // Internal API
         public static final String UI_CASE_VIEW = "application/vnd.uk.gov.hmcts.ccd-data-store-api.ui-case-view.v2+json;charset=UTF-8";
+
+        // External API
+        public static final String START_TRIGGER = "application/vnd.uk.gov.hmcts.ccd-data-store-api.start-trigger.v2+json;charset=UTF-8";
```
Comment (vlaurin) [inline src/main/java/uk/gov/hmcts/ccd/v2/V2.java]: There's already an *External API* section above, this line could be added to it.

File: src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/UserProfile.java [hunk 4]
```diff
@@ -44,4 +45,8 @@ public void setDefaultSettings(DefaultSettings defaultSettings) {
         this.defaultSettings = defaultSettings;
     }
 
+    @JsonIgnore
+    public String getUserId() {
+        return user.getIdamProperties().getId();
+    }
```
Comment (vlaurin) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/UserProfile.java]: Is this change needed?

File: src/main/java/uk/gov/hmcts/ccd/v2/V2.java (hunk 3 above)
Comment (andrewfolga) [reply src/main/java/uk/gov/hmcts/ccd/v2/V2.java]: Done

File: src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/UserProfile.java (hunk 4 above)
Comment (andrewfolga) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/UserProfile.java]: Done

File: src/main/java/uk/gov/hmcts/ccd/domain/service/startevent/DefaultStartEventOperation.java (hunk 1 above)
Comment (andrewfolga) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/startevent/DefaultStartEventOperation.java]: Good point

File: src/aat/java/uk/gov/hmcts/ccd/datastore/tests/v2/external/GetStartTriggerTest.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+import static uk.gov.hmcts.ccd.datastore.tests.fixture.AATCaseType.Event.CREATE;
+import static uk.gov.hmcts.ccd.datastore.tests.fixture.AATCaseType.JURISDICTION;
+
+@DisplayName("Get start trigger by case type and event ids")
+class GetStartTriggerTest extends BaseTest {
+    private static final JsonNodeFactory JSON_NODE_FACTORY = new JsonNodeFactory(false);
+    private static final ObjectNode EMPTY_OBJECT = JSON_NODE_FACTORY.objectNode();
+    private static final String EMPTY_OBJECT_STRING = EMPTY_OBJECT.toString();
+    private static final String INVALID_CASE_TYPE_ID = "unkonwnCaseType";
+    private static final String INVALID_EVENT_TRIGGER_ID = "unkonwnEvent";
```
[the remark this answers is not in the document]
Comment (andrewfolga) [reply src/aat/java/uk/gov/hmcts/ccd/datastore/tests/v2/external/GetStartTriggerTest.java]: :)

File: src/main/java/uk/gov/hmcts/ccd/domain/service/startevent/DefaultStartEventOperation.java (hunk 2 above)
Comment (andrewfolga) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/startevent/DefaultStartEventOperation.java]: Leftover from previous approach. Will cleanup.

File: src/main/java/uk/gov/hmcts/ccd/config/SwaggerConfiguration.java [hunk 6]
```diff
@@ -53,6 +54,7 @@ public Docket apiV2() {
             .groupName("v2")
             .select()
             .apis(RequestHandlerSelectors.basePackage(CaseController.class.getPackage().getName()))
+            .apis(RequestHandlerSelectors.basePackage(StartTriggerController.class.getPackage().getName()))
```
Comment (vlaurin) [inline src/main/java/uk/gov/hmcts/ccd/config/SwaggerConfiguration.java]: This is not needed as this is per package, and the package of `StartTriggerController.class` is the same as `CaseController.class`.

File: src/main/java/uk/gov/hmcts/ccd/domain/service/startevent/ClassifiedStartEventOperation.java [hunk 7]
```diff
@@ -35,6 +35,13 @@ public ClassifiedStartEventOperation(@Qualifier("default") StartEventOperation s
...
+    @Override
+    public StartEventTrigger triggerStartForCaseType(String caseTypeId, String eventTriggerId, Boolean ignoreWarning) {
+        return startEventOperation.triggerStartForCaseType(caseTypeId,
+                                                           eventTriggerId,
+                                                           ignoreWarning);
+    }
+
     @Override
     public StartEventTrigger triggerStartForCaseType(String uid, String jurisdictionId, String caseTypeId, String eventTriggerId, Boolean ignoreWarning) {
         return startEventOperation.triggerStartForCaseType(uid,
```
Comment (vlaurin) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/startevent/ClassifiedStartEventOperation.java]: Could you just replace that code by
```
startEventOperation.triggerStartForCaseType(caseTypeId,
                                                           eventTriggerId,
                                                           ignoreWarning);
```
